### DISCUSSION TITLE
About porting in a controller from Crazyflie_nmpc

### DISCUSSION THREAD

**Discussion by Yiqing-Gu** [gsilano/CrazyS#94] [Q&A]
Hi,

I am currently trying to port in a controller from crazyflie_nmpc to CrazyS. 

Now, I am looking at controller.cpp in nmpc (https://github.com/bcbarbara/crazyflie_nmpc/blob/master/crazyflie_controller/src/controller.cpp).

However, it is my first time to port in a controller. I wonder whether I am on the right track or anything suggestions for me?

Thanks.

**Comment by gsilano** [comment]
Well, if I were you, the first step would be to understand the control chain and the control signals involved. You could start reading the two conference proceedings attached to the repository. You can also use the repository's GitHub issue tracker to learn more. Simply, open a thread if you need any kind of assistance (only serious questions and issues).

Not so sure, but if I'm not mistaken, the controller was made to control a real Crazyflie quad-rotor. In that case, the control signals should be limited in the range [-1,1] for the roll, pitch and yaw rate and between [0,1] for the trust. Therefore, these values ​​should be mapped to work in CrazyS where we assume we are able to control  the propeller angular velocities ranging between [0, 2300] rad/s (check, I don't remember the correct values). However, the control signals are normalized quantities that are easy to remap. Also, you should understand the frequency each block works. This could affect the stability of the system.

In summary, here's how I would approach the problem:

1. I would read the papers related to the repository. From here, I would try to sketch up the control chain, detailing the signals and frequencies involved. To each block I would associate the corresponding cpp file, so as to create a map between paper and code.
2. I would understand what are the functions in CrazyS, how information is exchanged, from waypoint generation to vehicle control. Here too, a chain of control, signals and respective frequencies would help a lot.
3. I would try to understand if the controller can be taken as it is, by simply changing the parameters, or if it needs further modifications or an upstream or downstream interface block.
4. etc

I hope this helps. Please, keep me posted on the updated. If I can, I will try to help.

**Reply by Yiqing-Gu** [reply]
Thanks for your suggestions.
I have been working with crazyflie_nmpc for a while. My current plan is using ROS commands to monitor which topics nmpc subscribes or publishes. However, nmpc does not work well in my system. The last step in the installation always gave me errors. I checked issues and the problem still does exist now. What's more, I found there are a lot of answers haven't been answered for a long time. I guess this repository is more likely to be deprecated?
I saw you mentioned Lee's geometric controller in another question post. I read its paper and I found it is really similar to what I am doing right now. Could you tell me whether Lee's geometric controller is also a good choice for me? If so, is there something else I need to know since you are also working on this controller? I saw you said Lee's geometric controller needs to be tuned.

**Reply by gsilano** [reply]
I'm not sure if the repository is deprecated, but it could be that the author doesn't have time to reply to all messages. Try to tell me what's wrong with the installation. There may be a possibility that I can help in some way. Of course, I'd appreciate it if I could port the controller to CrazyS :D 
Yes, it needs to be tuned. This is the [MATLAB code of Lee's controller implementation available online](https://github.com/plusk01/se3quad), and this is the link to the work done by Stefano Grossi ([code](https://github.com/sgrossi/estensione_crazys), [Wiki](https://github.com/gsilano/CrazyS/wiki/Interfacing-CrazyS-through-MATLAB), [my open pull request](https://github.com/gsilano/CrazyS/pull/86)).

**Reply by Yiqing-Gu** [reply]
In crazyflie_nmpc, I feel confused with which acados folder I should put t_renderer in because I have two "acados". And I did not find a original "bin" folder. I created one for each of "acados" and put t_renderer (version got stripped) in them. No matter which acados I added to path. I got same error about no t_renderer.

This is the command I used to add a path: "export ACADOS_SOURCE_DIR=/home/cas-lab/catkin_ws/src/crazyflie_nmpc/acados"

The following is the error message:

/usr/lib/python3/dist-packages/requests/__init__.py:89: RequestsDependencyWarning: urllib3 (1.26.9) or chardet (3.0.4) doesn't match a supported version!
warnings.warn("urllib3 ({}) or chardet ({}) doesn't match a supported "
Traceback (most recent call last):
  File "generate_c_code.py", line 163, in <module>
    acados_solver = generate_solver(ra, json_file = 'acados_ocp.json')
  File "/home/cas-lab/.local/lib/python3.8/site-packages/acados_template/generate_solver.py", line 55, in generate_solver
    raise Exception(
Exception: /home/cas-lab/.local/lib/python3.8/site-packages/acados_template/../../..//bin/t_renderer Not found.
In order to be able to successfully render C code templates,you need to download the t_renderer binaries for your platform from https://github.com/acados/tera_renderer/releases/ and place them in <acados_root>/bin (please strip the version and platform from the binaries e.g

**Reply by gsilano** [reply]
I see your point. Have you tried to open an issue on the ACADOS official repository (https://github.com/acados/acados)?

**Reply by Yiqing-Gu** [reply]
Thanks for any supports you have offered again. I really would like to help you import nmpc as a part of CrazyS controllers. However, I was assigned a task to make CrazyS work with a trajectory generator as soon as possible. In the issue #92, you helped me figure out I needed a tracking controller to let a trajectory generator work in CrazyS. Just out of curiosity, whether Lee's geometric controller can satisfy my requirement? According to its paper, it should corporate with a tracking controller. 
I cloned Lee's geometric controller and I will try to boot simulation to check what kinds of bugs I will have.

**Reply by Yiqing-Gu** [reply]
> I see your point. Have you tried to open an issue on the ACADOS official repository (https://github.com/acados/acados)?

It is a good point. I haven't tried that yet. I guess the "acados" used in nmpc is not the latest and it caused some issues?

**Reply by gsilano** [reply]
Yes, this is what you need. You can find Lee's controller implementation in RotorS as well: https://github.com/ethz-asl/rotors_simulator/blob/master/rotors_control/src/nodes/lee_position_controller_node.cpp

**Reply by gsilano** [reply]
> > I see your point. Have you tried to open an issue on the ACADOS official repository (https://github.com/acados/acados)?
> 
> It is a good point. I haven't tried that yet. I guess the "acados" used in nmpc is not the latest and it caused some issues?

Maybe I do not know. It does not matter. You can implement Lee's tracking controller or NMPC. Now, as with Lee, you can use the MATLAB and RotorS implementation that I pointed out to you, and Stefano's thesis for the parameters (he took a long time to set them up).

**Reply by Yiqing-Gu** [reply]
Thanks for your useful material. I am gonna check those implementations to figure out where I should start and what I can do right now.

**Reply by Yiqing-Gu** [reply]
Hi, I have been working on Lee's controller for a while. I am using Ubuntu 20.04 and ROS noetic right now. I found package contains Lee's controller did not work well with Ubuntu 20.04 and ROS noetic? I got a compilation error while compiling this package. And it was as same as the one I had when I used melodic CrazyS in 20.04.
So I am using noetic CrazyS and I imported all unique files for Lee's controller to noetic CrazyS package. Compilation was successful. But now I got an error when running launch file about Lee's controller which told me

"ERROR: cannot launch node of type [rotors_control/lee_position_controller_alpha_node]: Cannot locate node of type [lee_position_controller_alpha_node] in package [rotors_control]. Make sure file exists in package path and permission is set to executable (chmod +x)
ERROR: cannot launch node of type [rotors_gazebo/hovering_lee]: Cannot locate node of type [hovering_lee] in package [rotors_gazebo]. Make sure file exists in package path and permission is set to executable (chmod +x)"

I confirmed that these two files were moved from the old package and set executable. However, I still could not git rid of errors. Could you give me some suggestions?

**Reply by gsilano** [reply]
What package are you referring to for Lee's controller?
As for the build issues, it's weird to me as GitHub Action has no issues. See the recent compilation report: https://github.com/gsilano/CrazyS/actions/runs/2605230058. Here are the ones for Melodic: https://github.com/gsilano/CrazyS/actions/runs/2605238411

I think the launch file is not well sourced in the bashrc. Could you send the contents of this file here?

**Reply by Yiqing-Gu** [reply]
> What package are you referring to for Lee's controller? As for the build issues, it's weird to me as GitHub Action has no issues. See the recent compilation report: https://github.com/gsilano/CrazyS/actions/runs/2605230058. Here are the ones for Melodic: https://github.com/gsilano/CrazyS/actions/runs/2605238411
> 
> I think the launch file is not well sourced in the bashrc. Could you send the contents of this file here?

Apologize for misunderstanding. 
Lee's controller I meant was branch "pull/lee_controller" in CrazyS. I cloned "pull/lee_controller" and tried to build it but I got some errors under ROS noetic. I guess "pull/lee_controller" does not fully support ROS noetic since it was released much earlier than the noetic branch?

The launch file I used was "[crazyflie_hovering_lee_controller_alpha.launch](https://github.com/gsilano/CrazyS/blob/pull/lee_controller/rotors_gazebo/launch/crazyflie_hovering_lee_controller_alpha.launch)". Since I moved Lee's controller components to noetic package directly, it can be something wrong with my operation.

**Reply by gsilano** [reply]
Now I see your point. Well, Lee's controller-related C++ files should be compatible with every version of ROS. However, the pull request cannot be copied and pasted from one release of CrazyS to another. Requires minimal customization. Then, you could clone the appropriate CrazyS release for your ROS version, and integrate the changes from there. Please note that this is a draft, so it may not compile or work correctly. It is definitely a great starting point, and I would be happy if you could finish my job.

**Reply by Yiqing-Gu** [reply]
Hi, Lee's controller package can be successfully compiled under ROS noetic right now. I tweaked two files in plugins and revised default python configurations. 
Now after executing "crazyflie_hovering_lee_controller_alpha.launch", I found every component in simulation could be loaded. However, crazyflie always stayed on the ground but propellers were spining. I would like to make sure whether it was the bug you had before? Or it was just related to my wrong configuration?
Thanks.

**Reply by gsilano** [reply]
Yes, it's correct. This is why I haven't merged the pull request yet. You need to complete the code and do some debugging.
